upstream-sha256: d50e91f4ee8efa2b944a38768f26fd772439f293392ea1e09574bee55fab673e
actual-sha256: 055d5978c253581fd5730cf11a901231c52f62ad15179b37d8a26684f9236fca
--- upstream.expected
+++ tolk.actual
@@ -48,91 +48,109 @@
    8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
    9 Ops.SHRINK          :            dtypes.float                             [6, 2, 8]                        None
   10 Ops.LOAD            :            dtypes.float                             [9]                              None
-  11 Ops.INDEX           :            dtypes.float                             [10, 2]                          None
-  12 Ops.CMPNE           :            dtypes.bool                              [11, 11]                         None
-  13 Ops.CONST           :            dtypes.bool                              []                               True
-  14 Ops.CAST            :            dtypes.bool                              ['True']                         dtypes.bool
-  15 Ops.XOR             :            dtypes.bool                              [12, 14]                         None
-  16 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
-  17 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
-  18 Ops.WHERE           :            dtypes.float                             [15, 11, 17]                     None
-  19 Ops.CONST           :            dtypes.weakint                           []                               1
-  20 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-  21 Ops.INDEX           :            dtypes.float                             [10, 20]                         None
+  11 Ops.CONST           :            dtypes.weakint                           []                               3
+  12 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
+  13 Ops.INDEX           :            dtypes.float                             [10, 12]                         None
+  14 Ops.CMPNE           :            dtypes.bool                              [13, 13]                         None
+  15 Ops.CONST           :            dtypes.bool                              []                               True
+  16 Ops.CAST            :            dtypes.bool                              ['True']                         dtypes.bool
+  17 Ops.XOR             :            dtypes.bool                              [14, 16]                         None
+  18 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
+  19 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
+  20 Ops.WHERE           :            dtypes.float                             [17, 13, 19]                     None
+  21 Ops.INDEX           :            dtypes.float                             [10, 4]                          None
   22 Ops.CMPNE           :            dtypes.bool                              [21, 21]                         None
-  23 Ops.XOR             :            dtypes.bool                              [22, 14]                         None
-  24 Ops.WHERE           :            dtypes.float                             [23, 21, 17]                     None
-  25 Ops.CMPLT           :            dtypes.bool                              [18, 24]                         None
-  26 Ops.WHERE           :            dtypes.float                             [25, 24, 18]                     None
-  27 Ops.INDEX           :            dtypes.float                             [10, 4]                          None
+  23 Ops.XOR             :            dtypes.bool                              [22, 16]                         None
+  24 Ops.WHERE           :            dtypes.float                             [23, 21, 19]                     None
+  25 Ops.CONST           :            dtypes.weakint                           []                               1
+  26 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  27 Ops.INDEX           :            dtypes.float                             [10, 26]                         None
   28 Ops.CMPNE           :            dtypes.bool                              [27, 27]                         None
-  29 Ops.XOR             :            dtypes.bool                              [28, 14]                         None
-  30 Ops.WHERE           :            dtypes.float                             [29, 27, 17]                     None
-  31 Ops.CMPLT           :            dtypes.bool                              [26, 30]                         None
-  32 Ops.WHERE           :            dtypes.float                             [31, 30, 26]                     None
-  33 Ops.CONST           :            dtypes.weakint                           []                               3
-  34 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
-  35 Ops.INDEX           :            dtypes.float                             [10, 34]                         None
-  36 Ops.CMPNE           :            dtypes.bool                              [35, 35]                         None
-  37 Ops.XOR             :            dtypes.bool                              [36, 14]                         None
-  38 Ops.WHERE           :            dtypes.float                             [37, 35, 17]                     None
-  39 Ops.CMPLT           :            dtypes.bool                              [32, 38]                         None
-  40 Ops.WHERE           :            dtypes.float                             [39, 38, 32]                     None
-  41 Ops.SHRINK          :            dtypes.float                             [6, 8, 8]                        None
-  42 Ops.LOAD            :            dtypes.float                             [41]                             None
-  43 Ops.INDEX           :            dtypes.float                             [42, 2]                          None
-  44 Ops.CMPNE           :            dtypes.bool                              [43, 43]                         None
-  45 Ops.XOR             :            dtypes.bool                              [44, 14]                         None
-  46 Ops.WHERE           :            dtypes.float                             [45, 43, 17]                     None
-  47 Ops.INDEX           :            dtypes.float                             [42, 20]                         None
-  48 Ops.CMPNE           :            dtypes.bool                              [47, 47]                         None
-  49 Ops.XOR             :            dtypes.bool                              [48, 14]                         None
-  50 Ops.WHERE           :            dtypes.float                             [49, 47, 17]                     None
-  51 Ops.CMPLT           :            dtypes.bool                              [46, 50]                         None
-  52 Ops.WHERE           :            dtypes.float                             [51, 50, 46]                     None
-  53 Ops.INDEX           :            dtypes.float                             [42, 4]                          None
+  29 Ops.XOR             :            dtypes.bool                              [28, 16]                         None
+  30 Ops.WHERE           :            dtypes.float                             [29, 27, 19]                     None
+  31 Ops.INDEX           :            dtypes.float                             [10, 2]                          None
+  32 Ops.CMPNE           :            dtypes.bool                              [31, 31]                         None
+  33 Ops.XOR             :            dtypes.bool                              [32, 16]                         None
+  34 Ops.WHERE           :            dtypes.float                             [33, 31, 19]                     None
+  35 Ops.CMPLT           :            dtypes.bool                              [30, 34]                         None
+  36 Ops.CMPNE           :            dtypes.bool                              [34, 34]                         None
+  37 Ops.WHERE           :            dtypes.float                             [36, 34, 30]                     None
+  38 Ops.WHERE           :            dtypes.float                             [35, 34, 37]                     None
+  39 Ops.CMPLT           :            dtypes.bool                              [24, 38]                         None
+  40 Ops.CMPNE           :            dtypes.bool                              [38, 38]                         None
+  41 Ops.WHERE           :            dtypes.float                             [40, 38, 24]                     None
+  42 Ops.WHERE           :            dtypes.float                             [39, 38, 41]                     None
+  43 Ops.CMPLT           :            dtypes.bool                              [20, 42]                         None
+  44 Ops.CMPNE           :            dtypes.bool                              [42, 42]                         None
+  45 Ops.WHERE           :            dtypes.float                             [44, 42, 20]                     None
+  46 Ops.WHERE           :            dtypes.float                             [43, 42, 45]                     None
+  47 Ops.SHRINK          :            dtypes.float                             [6, 8, 8]                        None
+  48 Ops.LOAD            :            dtypes.float                             [47]                             None
+  49 Ops.INDEX           :            dtypes.float                             [48, 12]                         None
+  50 Ops.CMPNE           :            dtypes.bool                              [49, 49]                         None
+  51 Ops.XOR             :            dtypes.bool                              [50, 16]                         None
+  52 Ops.WHERE           :            dtypes.float                             [51, 49, 19]                     None
+  53 Ops.INDEX           :            dtypes.float                             [48, 4]                          None
   54 Ops.CMPNE           :            dtypes.bool                              [53, 53]                         None
-  55 Ops.XOR             :            dtypes.bool                              [54, 14]                         None
-  56 Ops.WHERE           :            dtypes.float                             [55, 53, 17]                     None
-  57 Ops.CMPLT           :            dtypes.bool                              [52, 56]                         None
-  58 Ops.WHERE           :            dtypes.float                             [57, 56, 52]                     None
-  59 Ops.INDEX           :            dtypes.float                             [42, 34]                         None
-  60 Ops.CMPNE           :            dtypes.bool                              [59, 59]                         None
-  61 Ops.XOR             :            dtypes.bool                              [60, 14]                         None
-  62 Ops.WHERE           :            dtypes.float                             [61, 59, 17]                     None
-  63 Ops.CMPLT           :            dtypes.bool                              [58, 62]                         None
-  64 Ops.WHERE           :            dtypes.float                             [63, 62, 58]                     None
-  65 Ops.STACK           :            dtypes.float                             [40, 64]                         None
-  66 Ops.STORE           :            dtypes.void                              [5, 65]                          None
-  67 Ops.INDEX           :            dtypes.float                             [0, 4]                           None
-  68 Ops.CONST           :            dtypes.weakint                           []                               8
-  69 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-  70 Ops.SHRINK          :            dtypes.float                             [6, 69, 8]                       None
-  71 Ops.LOAD            :            dtypes.float                             [70]                             None
-  72 Ops.INDEX           :            dtypes.float                             [71, 2]                          None
-  73 Ops.CMPNE           :            dtypes.bool                              [72, 72]                         None
-  74 Ops.XOR             :            dtypes.bool                              [73, 14]                         None
-  75 Ops.WHERE           :            dtypes.float                             [74, 72, 17]                     None
-  76 Ops.INDEX           :            dtypes.float                             [71, 20]                         None
-  77 Ops.CMPNE           :            dtypes.bool                              [76, 76]                         None
-  78 Ops.XOR             :            dtypes.bool                              [77, 14]                         None
-  79 Ops.WHERE           :            dtypes.float                             [78, 76, 17]                     None
-  80 Ops.CMPLT           :            dtypes.bool                              [75, 79]                         None
-  81 Ops.WHERE           :            dtypes.float                             [80, 79, 75]                     None
-  82 Ops.INDEX           :            dtypes.float                             [71, 4]                          None
-  83 Ops.CMPNE           :            dtypes.bool                              [82, 82]                         None
-  84 Ops.XOR             :            dtypes.bool                              [83, 14]                         None
-  85 Ops.WHERE           :            dtypes.float                             [84, 82, 17]                     None
-  86 Ops.CMPLT           :            dtypes.bool                              [81, 85]                         None
-  87 Ops.WHERE           :            dtypes.float                             [86, 85, 81]                     None
-  88 Ops.INDEX           :            dtypes.float                             [71, 34]                         None
+  55 Ops.XOR             :            dtypes.bool                              [54, 16]                         None
+  56 Ops.WHERE           :            dtypes.float                             [55, 53, 19]                     None
+  57 Ops.INDEX           :            dtypes.float                             [48, 26]                         None
+  58 Ops.CMPNE           :            dtypes.bool                              [57, 57]                         None
+  59 Ops.XOR             :            dtypes.bool                              [58, 16]                         None
+  60 Ops.WHERE           :            dtypes.float                             [59, 57, 19]                     None
+  61 Ops.INDEX           :            dtypes.float                             [48, 2]                          None
+  62 Ops.CMPNE           :            dtypes.bool                              [61, 61]                         None
+  63 Ops.XOR             :            dtypes.bool                              [62, 16]                         None
+  64 Ops.WHERE           :            dtypes.float                             [63, 61, 19]                     None
+  65 Ops.CMPLT           :            dtypes.bool                              [60, 64]                         None
+  66 Ops.CMPNE           :            dtypes.bool                              [64, 64]                         None
+  67 Ops.WHERE           :            dtypes.float                             [66, 64, 60]                     None
+  68 Ops.WHERE           :            dtypes.float                             [65, 64, 67]                     None
+  69 Ops.CMPLT           :            dtypes.bool                              [56, 68]                         None
+  70 Ops.CMPNE           :            dtypes.bool                              [68, 68]                         None
+  71 Ops.WHERE           :            dtypes.float                             [70, 68, 56]                     None
+  72 Ops.WHERE           :            dtypes.float                             [69, 68, 71]                     None
+  73 Ops.CMPLT           :            dtypes.bool                              [52, 72]                         None
+  74 Ops.CMPNE           :            dtypes.bool                              [72, 72]                         None
+  75 Ops.WHERE           :            dtypes.float                             [74, 72, 52]                     None
+  76 Ops.WHERE           :            dtypes.float                             [73, 72, 75]                     None
+  77 Ops.STACK           :            dtypes.float                             [46, 76]                         None
+  78 Ops.STORE           :            dtypes.void                              [5, 77]                          None
+  79 Ops.INDEX           :            dtypes.float                             [0, 4]                           None
+  80 Ops.CONST           :            dtypes.weakint                           []                               8
+  81 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+  82 Ops.SHRINK          :            dtypes.float                             [6, 81, 8]                       None
+  83 Ops.LOAD            :            dtypes.float                             [82]                             None
+  84 Ops.INDEX           :            dtypes.float                             [83, 12]                         None
+  85 Ops.CMPNE           :            dtypes.bool                              [84, 84]                         None
+  86 Ops.XOR             :            dtypes.bool                              [85, 16]                         None
+  87 Ops.WHERE           :            dtypes.float                             [86, 84, 19]                     None
+  88 Ops.INDEX           :            dtypes.float                             [83, 4]                          None
   89 Ops.CMPNE           :            dtypes.bool                              [88, 88]                         None
-  90 Ops.XOR             :            dtypes.bool                              [89, 14]                         None
-  91 Ops.WHERE           :            dtypes.float                             [90, 88, 17]                     None
-  92 Ops.CMPLT           :            dtypes.bool                              [87, 91]                         None
-  93 Ops.WHERE           :            dtypes.float                             [92, 91, 87]                     None
-  94 Ops.STORE           :            dtypes.void                              [67, 93]                         None
-  95 Ops.SINK            :            dtypes.void                              [66, 66, 94]                     KernelInfo(name='r_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL))), opts_to_apply=None, estimates=None, beam=0)
+  90 Ops.XOR             :            dtypes.bool                              [89, 16]                         None
+  91 Ops.WHERE           :            dtypes.float                             [90, 88, 19]                     None
+  92 Ops.INDEX           :            dtypes.float                             [83, 26]                         None
+  93 Ops.CMPNE           :            dtypes.bool                              [92, 92]                         None
+  94 Ops.XOR             :            dtypes.bool                              [93, 16]                         None
+  95 Ops.WHERE           :            dtypes.float                             [94, 92, 19]                     None
+  96 Ops.INDEX           :            dtypes.float                             [83, 2]                          None
+  97 Ops.CMPNE           :            dtypes.bool                              [96, 96]                         None
+  98 Ops.XOR             :            dtypes.bool                              [97, 16]                         None
+  99 Ops.WHERE           :            dtypes.float                             [98, 96, 19]                     None
+ 100 Ops.CMPLT           :            dtypes.bool                              [95, 99]                         None
+ 101 Ops.CMPNE           :            dtypes.bool                              [99, 99]                         None
+ 102 Ops.WHERE           :            dtypes.float                             [101, 99, 95]                    None
+ 103 Ops.WHERE           :            dtypes.float                             [100, 99, 102]                   None
+ 104 Ops.CMPLT           :            dtypes.bool                              [91, 103]                        None
+ 105 Ops.CMPNE           :            dtypes.bool                              [103, 103]                       None
+ 106 Ops.WHERE           :            dtypes.float                             [105, 103, 91]                   None
+ 107 Ops.WHERE           :            dtypes.float                             [104, 103, 106]                  None
+ 108 Ops.CMPLT           :            dtypes.bool                              [87, 107]                        None
+ 109 Ops.CMPNE           :            dtypes.bool                              [107, 107]                       None
+ 110 Ops.WHERE           :            dtypes.float                             [109, 107, 87]                   None
+ 111 Ops.WHERE           :            dtypes.float                             [108, 107, 110]                  None
+ 112 Ops.STORE           :            dtypes.void                              [79, 111]                        None
+ 113 Ops.SINK            :            dtypes.void                              [78, 78, 112]                    KernelInfo(name='r_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 2 ===
    0 Ops.PARAM           :            dtypes.int                               []                               ParamArg(0, dtypes.int, 3, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               0
@@ -534,129 +552,147 @@
    8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
    9 Ops.SHRINK          :            dtypes.int                               [6, 2, 8]                        None
   10 Ops.LOAD            :            dtypes.int                               [9]                              None
-  11 Ops.INDEX           :            dtypes.int                               [10, 2]                          None
-  12 Ops.PARAM           :            dtypes.int                               []                               ParamArg(2, dtypes.int, 3, device='CPU')
-  13 Ops.SHRINK          :            dtypes.int                               [12, 2, 4]                       None
-  14 Ops.LOAD            :            dtypes.int                               [13]                             None
-  15 Ops.INDEX           :            dtypes.int                               [14, 2]                          None
-  16 Ops.CMPNE           :            dtypes.bool                              [11, 15]                         None
-  17 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 12, device='CPU')
-  18 Ops.SHRINK          :            dtypes.float                             [17, 2, 8]                       None
-  19 Ops.LOAD            :            dtypes.float                             [18]                             None
-  20 Ops.INDEX           :            dtypes.float                             [19, 2]                          None
-  21 Ops.CMPNE           :            dtypes.bool                              [20, 20]                         None
-  22 Ops.CONST           :            dtypes.bool                              []                               True
-  23 Ops.CAST            :            dtypes.bool                              ['True']                         dtypes.bool
-  24 Ops.XOR             :            dtypes.bool                              [21, 23]                         None
-  25 Ops.AND             :            dtypes.bool                              [16, 24]                         None
-  26 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
-  27 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
-  28 Ops.WHERE           :            dtypes.float                             [25, 20, 27]                     None
-  29 Ops.CONST           :            dtypes.weakint                           []                               1
-  30 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-  31 Ops.INDEX           :            dtypes.int                               [10, 30]                         None
-  32 Ops.CMPNE           :            dtypes.bool                              [31, 15]                         None
-  33 Ops.INDEX           :            dtypes.float                             [19, 30]                         None
+  11 Ops.CONST           :            dtypes.weakint                           []                               3
+  12 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
+  13 Ops.INDEX           :            dtypes.int                               [10, 12]                         None
+  14 Ops.PARAM           :            dtypes.int                               []                               ParamArg(2, dtypes.int, 3, device='CPU')
+  15 Ops.SHRINK          :            dtypes.int                               [14, 2, 4]                       None
+  16 Ops.LOAD            :            dtypes.int                               [15]                             None
+  17 Ops.INDEX           :            dtypes.int                               [16, 2]                          None
+  18 Ops.CMPNE           :            dtypes.bool                              [13, 17]                         None
+  19 Ops.PARAM           :            dtypes.float                             []                               ParamArg(3, dtypes.float, 12, device='CPU')
+  20 Ops.SHRINK          :            dtypes.float                             [19, 2, 8]                       None
+  21 Ops.LOAD            :            dtypes.float                             [20]                             None
+  22 Ops.INDEX           :            dtypes.float                             [21, 12]                         None
+  23 Ops.CMPNE           :            dtypes.bool                              [22, 22]                         None
+  24 Ops.CONST           :            dtypes.bool                              []                               True
+  25 Ops.CAST            :            dtypes.bool                              ['True']                         dtypes.bool
+  26 Ops.XOR             :            dtypes.bool                              [23, 25]                         None
+  27 Ops.AND             :            dtypes.bool                              [18, 26]                         None
+  28 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
+  29 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
+  30 Ops.WHERE           :            dtypes.float                             [27, 22, 29]                     None
+  31 Ops.INDEX           :            dtypes.int                               [10, 4]                          None
+  32 Ops.CMPNE           :            dtypes.bool                              [31, 17]                         None
+  33 Ops.INDEX           :            dtypes.float                             [21, 4]                          None
   34 Ops.CMPNE           :            dtypes.bool                              [33, 33]                         None
-  35 Ops.XOR             :            dtypes.bool                              [34, 23]                         None
+  35 Ops.XOR             :            dtypes.bool                              [34, 25]                         None
   36 Ops.AND             :            dtypes.bool                              [32, 35]                         None
-  37 Ops.WHERE           :            dtypes.float                             [36, 33, 27]                     None
-  38 Ops.CMPLT           :            dtypes.bool                              [28, 37]                         None
-  39 Ops.WHERE           :            dtypes.float                             [38, 37, 28]                     None
-  40 Ops.INDEX           :            dtypes.int                               [10, 4]                          None
-  41 Ops.CMPNE           :            dtypes.bool                              [40, 15]                         None
-  42 Ops.INDEX           :            dtypes.float                             [19, 4]                          None
+  37 Ops.WHERE           :            dtypes.float                             [36, 33, 29]                     None
+  38 Ops.CONST           :            dtypes.weakint                           []                               1
+  39 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  40 Ops.INDEX           :            dtypes.int                               [10, 39]                         None
+  41 Ops.CMPNE           :            dtypes.bool                              [40, 17]                         None
+  42 Ops.INDEX           :            dtypes.float                             [21, 39]                         None
   43 Ops.CMPNE           :            dtypes.bool                              [42, 42]                         None
-  44 Ops.XOR             :            dtypes.bool                              [43, 23]                         None
+  44 Ops.XOR             :            dtypes.bool                              [43, 25]                         None
   45 Ops.AND             :            dtypes.bool                              [41, 44]                         None
-  46 Ops.WHERE           :            dtypes.float                             [45, 42, 27]                     None
-  47 Ops.CMPLT           :            dtypes.bool                              [39, 46]                         None
-  48 Ops.WHERE           :            dtypes.float                             [47, 46, 39]                     None
-  49 Ops.CONST           :            dtypes.weakint                           []                               3
-  50 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
-  51 Ops.INDEX           :            dtypes.int                               [10, 50]                         None
-  52 Ops.CMPNE           :            dtypes.bool                              [51, 15]                         None
-  53 Ops.INDEX           :            dtypes.float                             [19, 50]                         None
-  54 Ops.CMPNE           :            dtypes.bool                              [53, 53]                         None
-  55 Ops.XOR             :            dtypes.bool                              [54, 23]                         None
-  56 Ops.AND             :            dtypes.bool                              [52, 55]                         None
-  57 Ops.WHERE           :            dtypes.float                             [56, 53, 27]                     None
-  58 Ops.CMPLT           :            dtypes.bool                              [48, 57]                         None
-  59 Ops.WHERE           :            dtypes.float                             [58, 57, 48]                     None
-  60 Ops.INDEX           :            dtypes.int                               [14, 30]                         None
-  61 Ops.CMPNE           :            dtypes.bool                              [11, 60]                         None
-  62 Ops.SHRINK          :            dtypes.float                             [17, 8, 8]                       None
-  63 Ops.LOAD            :            dtypes.float                             [62]                             None
-  64 Ops.INDEX           :            dtypes.float                             [63, 2]                          None
-  65 Ops.CMPNE           :            dtypes.bool                              [64, 64]                         None
-  66 Ops.XOR             :            dtypes.bool                              [65, 23]                         None
-  67 Ops.AND             :            dtypes.bool                              [61, 66]                         None
-  68 Ops.WHERE           :            dtypes.float                             [67, 64, 27]                     None
-  69 Ops.CMPNE           :            dtypes.bool                              [31, 60]                         None
-  70 Ops.INDEX           :            dtypes.float                             [63, 30]                         None
+  46 Ops.WHERE           :            dtypes.float                             [45, 42, 29]                     None
+  47 Ops.INDEX           :            dtypes.int                               [10, 2]                          None
+  48 Ops.CMPNE           :            dtypes.bool                              [47, 17]                         None
+  49 Ops.INDEX           :            dtypes.float                             [21, 2]                          None
+  50 Ops.CMPNE           :            dtypes.bool                              [49, 49]                         None
+  51 Ops.XOR             :            dtypes.bool                              [50, 25]                         None
+  52 Ops.AND             :            dtypes.bool                              [48, 51]                         None
+  53 Ops.WHERE           :            dtypes.float                             [52, 49, 29]                     None
+  54 Ops.CMPLT           :            dtypes.bool                              [46, 53]                         None
+  55 Ops.CMPNE           :            dtypes.bool                              [53, 53]                         None
+  56 Ops.WHERE           :            dtypes.float                             [55, 53, 46]                     None
+  57 Ops.WHERE           :            dtypes.float                             [54, 53, 56]                     None
+  58 Ops.CMPLT           :            dtypes.bool                              [37, 57]                         None
+  59 Ops.CMPNE           :            dtypes.bool                              [57, 57]                         None
+  60 Ops.WHERE           :            dtypes.float                             [59, 57, 37]                     None
+  61 Ops.WHERE           :            dtypes.float                             [58, 57, 60]                     None
+  62 Ops.CMPLT           :            dtypes.bool                              [30, 61]                         None
+  63 Ops.CMPNE           :            dtypes.bool                              [61, 61]                         None
+  64 Ops.WHERE           :            dtypes.float                             [63, 61, 30]                     None
+  65 Ops.WHERE           :            dtypes.float                             [62, 61, 64]                     None
+  66 Ops.INDEX           :            dtypes.int                               [16, 39]                         None
+  67 Ops.CMPNE           :            dtypes.bool                              [13, 66]                         None
+  68 Ops.SHRINK          :            dtypes.float                             [19, 8, 8]                       None
+  69 Ops.LOAD            :            dtypes.float                             [68]                             None
+  70 Ops.INDEX           :            dtypes.float                             [69, 12]                         None
   71 Ops.CMPNE           :            dtypes.bool                              [70, 70]                         None
-  72 Ops.XOR             :            dtypes.bool                              [71, 23]                         None
-  73 Ops.AND             :            dtypes.bool                              [69, 72]                         None
-  74 Ops.WHERE           :            dtypes.float                             [73, 70, 27]                     None
-  75 Ops.CMPLT           :            dtypes.bool                              [68, 74]                         None
-  76 Ops.WHERE           :            dtypes.float                             [75, 74, 68]                     None
-  77 Ops.CMPNE           :            dtypes.bool                              [40, 60]                         None
-  78 Ops.INDEX           :            dtypes.float                             [63, 4]                          None
-  79 Ops.CMPNE           :            dtypes.bool                              [78, 78]                         None
-  80 Ops.XOR             :            dtypes.bool                              [79, 23]                         None
-  81 Ops.AND             :            dtypes.bool                              [77, 80]                         None
-  82 Ops.WHERE           :            dtypes.float                             [81, 78, 27]                     None
-  83 Ops.CMPLT           :            dtypes.bool                              [76, 82]                         None
-  84 Ops.WHERE           :            dtypes.float                             [83, 82, 76]                     None
-  85 Ops.CMPNE           :            dtypes.bool                              [51, 60]                         None
-  86 Ops.INDEX           :            dtypes.float                             [63, 50]                         None
-  87 Ops.CMPNE           :            dtypes.bool                              [86, 86]                         None
-  88 Ops.XOR             :            dtypes.bool                              [87, 23]                         None
-  89 Ops.AND             :            dtypes.bool                              [85, 88]                         None
-  90 Ops.WHERE           :            dtypes.float                             [89, 86, 27]                     None
-  91 Ops.CMPLT           :            dtypes.bool                              [84, 90]                         None
-  92 Ops.WHERE           :            dtypes.float                             [91, 90, 84]                     None
-  93 Ops.STACK           :            dtypes.float                             [59, 92]                         None
-  94 Ops.STORE           :            dtypes.void                              [5, 93]                          None
-  95 Ops.INDEX           :            dtypes.float                             [0, 4]                           None
-  96 Ops.INDEX           :            dtypes.int                               [12, 4]                          None
-  97 Ops.LOAD            :            dtypes.int                               [96]                             None
-  98 Ops.CMPNE           :            dtypes.bool                              [11, 97]                         None
-  99 Ops.CONST           :            dtypes.weakint                           []                               8
- 100 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
- 101 Ops.SHRINK          :            dtypes.float                             [17, 100, 8]                     None
- 102 Ops.LOAD            :            dtypes.float                             [101]                            None
- 103 Ops.INDEX           :            dtypes.float                             [102, 2]                         None
- 104 Ops.CMPNE           :            dtypes.bool                              [103, 103]                       None
- 105 Ops.XOR             :            dtypes.bool                              [104, 23]                        None
- 106 Ops.AND             :            dtypes.bool                              [98, 105]                        None
- 107 Ops.WHERE           :            dtypes.float                             [106, 103, 27]                   None
- 108 Ops.CMPNE           :            dtypes.bool                              [31, 97]                         None
- 109 Ops.INDEX           :            dtypes.float                             [102, 30]                        None
- 110 Ops.CMPNE           :            dtypes.bool                              [109, 109]                       None
- 111 Ops.XOR             :            dtypes.bool                              [110, 23]                        None
- 112 Ops.AND             :            dtypes.bool                              [108, 111]                       None
- 113 Ops.WHERE           :            dtypes.float                             [112, 109, 27]                   None
- 114 Ops.CMPLT           :            dtypes.bool                              [107, 113]                       None
- 115 Ops.WHERE           :            dtypes.float                             [114, 113, 107]                  None
- 116 Ops.CMPNE           :            dtypes.bool                              [40, 97]                         None
- 117 Ops.INDEX           :            dtypes.float                             [102, 4]                         None
- 118 Ops.CMPNE           :            dtypes.bool                              [117, 117]                       None
- 119 Ops.XOR             :            dtypes.bool                              [118, 23]                        None
- 120 Ops.AND             :            dtypes.bool                              [116, 119]                       None
- 121 Ops.WHERE           :            dtypes.float                             [120, 117, 27]                   None
- 122 Ops.CMPLT           :            dtypes.bool                              [115, 121]                       None
- 123 Ops.WHERE           :            dtypes.float                             [122, 121, 115]                  None
- 124 Ops.CMPNE           :            dtypes.bool                              [51, 97]                         None
- 125 Ops.INDEX           :            dtypes.float                             [102, 50]                        None
- 126 Ops.CMPNE           :            dtypes.bool                              [125, 125]                       None
- 127 Ops.XOR             :            dtypes.bool                              [126, 23]                        None
- 128 Ops.AND             :            dtypes.bool                              [124, 127]                       None
- 129 Ops.WHERE           :            dtypes.float                             [128, 125, 27]                   None
- 130 Ops.CMPLT           :            dtypes.bool                              [123, 129]                       None
- 131 Ops.WHERE           :            dtypes.float                             [130, 129, 123]                  None
- 132 Ops.STORE           :            dtypes.void                              [95, 131]                        None
- 133 Ops.SINK            :            dtypes.void                              [94, 94, 132]                    KernelInfo(name='r_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL))), opts_to_apply=None, estimates=None, beam=0)
+  72 Ops.XOR             :            dtypes.bool                              [71, 25]                         None
+  73 Ops.AND             :            dtypes.bool                              [67, 72]                         None
+  74 Ops.WHERE           :            dtypes.float                             [73, 70, 29]                     None
+  75 Ops.CMPNE           :            dtypes.bool                              [31, 66]                         None
+  76 Ops.INDEX           :            dtypes.float                             [69, 4]                          None
+  77 Ops.CMPNE           :            dtypes.bool                              [76, 76]                         None
+  78 Ops.XOR             :            dtypes.bool                              [77, 25]                         None
+  79 Ops.AND             :            dtypes.bool                              [75, 78]                         None
+  80 Ops.WHERE           :            dtypes.float                             [79, 76, 29]                     None
+  81 Ops.CMPNE           :            dtypes.bool                              [40, 66]                         None
+  82 Ops.INDEX           :            dtypes.float                             [69, 39]                         None
+  83 Ops.CMPNE           :            dtypes.bool                              [82, 82]                         None
+  84 Ops.XOR             :            dtypes.bool                              [83, 25]                         None
+  85 Ops.AND             :            dtypes.bool                              [81, 84]                         None
+  86 Ops.WHERE           :            dtypes.float                             [85, 82, 29]                     None
+  87 Ops.CMPNE           :            dtypes.bool                              [47, 66]                         None
+  88 Ops.INDEX           :            dtypes.float                             [69, 2]                          None
+  89 Ops.CMPNE           :            dtypes.bool                              [88, 88]                         None
+  90 Ops.XOR             :            dtypes.bool                              [89, 25]                         None
+  91 Ops.AND             :            dtypes.bool                              [87, 90]                         None
+  92 Ops.WHERE           :            dtypes.float                             [91, 88, 29]                     None
+  93 Ops.CMPLT           :            dtypes.bool                              [86, 92]                         None
+  94 Ops.CMPNE           :            dtypes.bool                              [92, 92]                         None
+  95 Ops.WHERE           :            dtypes.float                             [94, 92, 86]                     None
+  96 Ops.WHERE           :            dtypes.float                             [93, 92, 95]                     None
+  97 Ops.CMPLT           :            dtypes.bool                              [80, 96]                         None
+  98 Ops.CMPNE           :            dtypes.bool                              [96, 96]                         None
+  99 Ops.WHERE           :            dtypes.float                             [98, 96, 80]                     None
+ 100 Ops.WHERE           :            dtypes.float                             [97, 96, 99]                     None
+ 101 Ops.CMPLT           :            dtypes.bool                              [74, 100]                        None
+ 102 Ops.CMPNE           :            dtypes.bool                              [100, 100]                       None
+ 103 Ops.WHERE           :            dtypes.float                             [102, 100, 74]                   None
+ 104 Ops.WHERE           :            dtypes.float                             [101, 100, 103]                  None
+ 105 Ops.STACK           :            dtypes.float                             [65, 104]                        None
+ 106 Ops.STORE           :            dtypes.void                              [5, 105]                         None
+ 107 Ops.INDEX           :            dtypes.float                             [0, 4]                           None
+ 108 Ops.INDEX           :            dtypes.int                               [14, 4]                          None
+ 109 Ops.LOAD            :            dtypes.int                               [108]                            None
+ 110 Ops.CMPNE           :            dtypes.bool                              [13, 109]                        None
+ 111 Ops.CONST           :            dtypes.weakint                           []                               8
+ 112 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+ 113 Ops.SHRINK          :            dtypes.float                             [19, 112, 8]                     None
+ 114 Ops.LOAD            :            dtypes.float                             [113]                            None
+ 115 Ops.INDEX           :            dtypes.float                             [114, 12]                        None
+ 116 Ops.CMPNE           :            dtypes.bool                              [115, 115]                       None
+ 117 Ops.XOR             :            dtypes.bool                              [116, 25]                        None
+ 118 Ops.AND             :            dtypes.bool                              [110, 117]                       None
+ 119 Ops.WHERE           :            dtypes.float                             [118, 115, 29]                   None
+ 120 Ops.CMPNE           :            dtypes.bool                              [31, 109]                        None
+ 121 Ops.INDEX           :            dtypes.float                             [114, 4]                         None
+ 122 Ops.CMPNE           :            dtypes.bool                              [121, 121]                       None
+ 123 Ops.XOR             :            dtypes.bool                              [122, 25]                        None
+ 124 Ops.AND             :            dtypes.bool                              [120, 123]                       None
+ 125 Ops.WHERE           :            dtypes.float                             [124, 121, 29]                   None
+ 126 Ops.CMPNE           :            dtypes.bool                              [40, 109]                        None
+ 127 Ops.INDEX           :            dtypes.float                             [114, 39]                        None
+ 128 Ops.CMPNE           :            dtypes.bool                              [127, 127]                       None
+ 129 Ops.XOR             :            dtypes.bool                              [128, 25]                        None
+ 130 Ops.AND             :            dtypes.bool                              [126, 129]                       None
+ 131 Ops.WHERE           :            dtypes.float                             [130, 127, 29]                   None
+ 132 Ops.CMPNE           :            dtypes.bool                              [47, 109]                        None
+ 133 Ops.INDEX           :            dtypes.float                             [114, 2]                         None
+ 134 Ops.CMPNE           :            dtypes.bool                              [133, 133]                       None
+ 135 Ops.XOR             :            dtypes.bool                              [134, 25]                        None
+ 136 Ops.AND             :            dtypes.bool                              [132, 135]                       None
+ 137 Ops.WHERE           :            dtypes.float                             [136, 133, 29]                   None
+ 138 Ops.CMPLT           :            dtypes.bool                              [131, 137]                       None
+ 139 Ops.CMPNE           :            dtypes.bool                              [137, 137]                       None
+ 140 Ops.WHERE           :            dtypes.float                             [139, 137, 131]                  None
+ 141 Ops.WHERE           :            dtypes.float                             [138, 137, 140]                  None
+ 142 Ops.CMPLT           :            dtypes.bool                              [125, 141]                       None
+ 143 Ops.CMPNE           :            dtypes.bool                              [141, 141]                       None
+ 144 Ops.WHERE           :            dtypes.float                             [143, 141, 125]                  None
+ 145 Ops.WHERE           :            dtypes.float                             [142, 141, 144]                  None
+ 146 Ops.CMPLT           :            dtypes.bool                              [119, 145]                       None
+ 147 Ops.CMPNE           :            dtypes.bool                              [145, 145]                       None
+ 148 Ops.WHERE           :            dtypes.float                             [147, 145, 119]                  None
+ 149 Ops.WHERE           :            dtypes.float                             [146, 145, 148]                  None
+ 150 Ops.STORE           :            dtypes.void                              [107, 149]                       None
+ 151 Ops.SINK            :            dtypes.void                              [106, 106, 150]                  KernelInfo(name='r_3_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(0, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 6 ===
    0 Ops.PARAM           :            dtypes.int                               []                               ParamArg(0, dtypes.int, 12, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               4
@@ -923,50 +959,52 @@
   10 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
   11 Ops.SHRINK          : 1          dtypes.float                             [5, 8, 10]                       None
   12 Ops.LOAD            : 1          dtypes.float                             [11]                             None
-  13 Ops.CONST           :            dtypes.weakint                           []                               0
-  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  15 Ops.INDEX           : 1          dtypes.float                             [12, 14]                         None
-  16 Ops.INDEX           : 1          dtypes.float                             [12, 7]                          None
-  17 Ops.CMPLT           : 1          dtypes.bool                              [15, 16]                         None
-  18 Ops.WHERE           : 1          dtypes.float                             [17, 16, 15]                     None
-  19 Ops.STORE           : 1          dtypes.void                              [4, 18]                          None
-  20 Ops.END             :            dtypes.void                              [19, 3]                          None
-  21 Ops.SINK            :            dtypes.void                              [20]                             KernelInfo(name='r_3_2', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)),), opts_to_apply=None, estimates=None, beam=0)
+  13 Ops.INDEX           : 1          dtypes.float                             [12, 7]                          None
+  14 Ops.CONST           :            dtypes.weakint                           []                               0
+  15 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  16 Ops.INDEX           : 1          dtypes.float                             [12, 15]                         None
+  17 Ops.CMPLT           : 1          dtypes.bool                              [13, 16]                         None
+  18 Ops.CMPNE           : 1          dtypes.bool                              [16, 16]                         None
+  19 Ops.WHERE           : 1          dtypes.float                             [18, 16, 13]                     None
+  20 Ops.WHERE           : 1          dtypes.float                             [17, 16, 19]                     None
+  21 Ops.STORE           : 1          dtypes.void                              [4, 20]                          None
+  22 Ops.END             :            dtypes.void                              [21, 3]                          None
+  23 Ops.SINK            :            dtypes.void                              [22]                             KernelInfo(name='r_3_2', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)),), opts_to_apply=None, estimates=None, beam=0)
 === kernel 12 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 3, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               3
    2 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
    3 Ops.RANGE           : 1          dtypes.int                               [2]                              (1, AxisType.WEAK)
    4 Ops.INDEX           : 1          dtypes.float                             [0, 3]                           None
-   5 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 6, device='CPU')
-   6 Ops.CONST           :            dtypes.weakint                           []                               1
-   7 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-   8 Ops.SHL             : 1          dtypes.int                               [3, 7]                           None
-   9 Ops.CONST           :            dtypes.weakint                           []                               2
-  10 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-  11 Ops.SHRINK          : 1          dtypes.float                             [5, 8, 10]                       None
-  12 Ops.LOAD            : 1          dtypes.float                             [11]                             None
-  13 Ops.CONST           :            dtypes.weakint                           []                               0
-  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  15 Ops.INDEX           : 1          dtypes.float                             [12, 14]                         None
-  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 3, device='CPU')
-  17 Ops.INDEX           : 1          dtypes.float                             [16, 3]                          None
-  18 Ops.LOAD            : 1          dtypes.float                             [17]                             None
-  19 Ops.SUB             : 1          dtypes.float                             [15, 18]                         None
-  20 Ops.CONST           :            dtypes.weakfloat                         []                               1.4426950216293335
-  21 Ops.CAST            :            dtypes.float                             ['1.4426950216293335']           dtypes.float
-  22 Ops.MUL             : 1          dtypes.float                             [19, 21]                         None
-  23 Ops.CMPNE           : 1          dtypes.bool                              [22, 22]                         None
-  24 Ops.CONST           :            dtypes.weakfloat                         []                               nan
-  25 Ops.CAST            :            dtypes.float                             ['nan']                          dtypes.float
-  26 Ops.CONST           :            dtypes.weakfloat                         []                               -150.0
-  27 Ops.CAST            :            dtypes.float                             ['-150.0']                       dtypes.float
-  28 Ops.CMPLT           : 1          dtypes.bool                              [22, 27]                         None
-  29 Ops.CONST           :            dtypes.weakfloat                         []                               0.0
-  30 Ops.CAST            :            dtypes.float                             ['0.0']                          dtypes.float
+   5 Ops.CONST           :            dtypes.weakfloat                         []                               0.0
+   6 Ops.CAST            :            dtypes.float                             ['0.0']                          dtypes.float
+   7 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 6, device='CPU')
+   8 Ops.CONST           :            dtypes.weakint                           []                               1
+   9 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  10 Ops.SHL             : 1          dtypes.int                               [3, 9]                           None
+  11 Ops.CONST           :            dtypes.weakint                           []                               2
+  12 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+  13 Ops.SHRINK          : 1          dtypes.float                             [7, 10, 12]                      None
+  14 Ops.LOAD            : 1          dtypes.float                             [13]                             None
+  15 Ops.CONST           :            dtypes.weakint                           []                               0
+  16 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  17 Ops.INDEX           : 1          dtypes.float                             [14, 16]                         None
+  18 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 3, device='CPU')
+  19 Ops.INDEX           : 1          dtypes.float                             [18, 3]                          None
+  20 Ops.LOAD            : 1          dtypes.float                             [19]                             None
+  21 Ops.SUB             : 1          dtypes.float                             [17, 20]                         None
+  22 Ops.CONST           :            dtypes.weakfloat                         []                               1.4426950216293335
+  23 Ops.CAST            :            dtypes.float                             ['1.4426950216293335']           dtypes.float
+  24 Ops.MUL             : 1          dtypes.float                             [21, 23]                         None
+  25 Ops.CMPNE           : 1          dtypes.bool                              [24, 24]                         None
+  26 Ops.CONST           :            dtypes.weakfloat                         []                               nan
+  27 Ops.CAST            :            dtypes.float                             ['nan']                          dtypes.float
+  28 Ops.CONST           :            dtypes.weakfloat                         []                               -150.0
+  29 Ops.CAST            :            dtypes.float                             ['-150.0']                       dtypes.float
+  30 Ops.CMPLT           : 1          dtypes.bool                              [24, 29]                         None
   31 Ops.CONST           :            dtypes.weakfloat                         []                               128.0
   32 Ops.CAST            :            dtypes.float                             ['128.0']                        dtypes.float
-  33 Ops.CMPLT           : 1          dtypes.bool                              [22, 32]                         None
+  33 Ops.CMPLT           : 1          dtypes.bool                              [24, 32]                         None
   34 Ops.CONST           :            dtypes.bool                              []                               True
   35 Ops.CAST            :            dtypes.bool                              ['True']                         dtypes.bool
   36 Ops.CMPNE           : 1          dtypes.bool                              [33, 35]                         None
@@ -974,14 +1012,14 @@
   38 Ops.CAST            :            dtypes.float                             ['inf']                          dtypes.float
   39 Ops.CONST           :            dtypes.weakfloat                         []                               0.00015359208919107914
   40 Ops.CAST            :            dtypes.float                             ['0.00015359208919107914']       dtypes.float
-  41 Ops.CMPNE           : 1          dtypes.bool                              [22, 38]                         None
+  41 Ops.CMPNE           : 1          dtypes.bool                              [24, 38]                         None
   42 Ops.CONST           :            dtypes.weakfloat                         []                               -inf
   43 Ops.CAST            :            dtypes.float                             ['-inf']                         dtypes.float
-  44 Ops.CMPNE           : 1          dtypes.bool                              [22, 43]                         None
-  45 Ops.WHERE           : 1          dtypes.float                             [44, 22, 30]                     None
-  46 Ops.WHERE           : 1          dtypes.float                             [23, 30, 45]                     None
-  47 Ops.WHERE           : 1          dtypes.float                             [41, 46, 30]                     None
-  48 Ops.CMPLT           : 1          dtypes.bool                              [47, 30]                         None
+  44 Ops.CMPNE           : 1          dtypes.bool                              [24, 43]                         None
+  45 Ops.WHERE           : 1          dtypes.float                             [44, 24, 6]                      None
+  46 Ops.WHERE           : 1          dtypes.float                             [25, 6, 45]                      None
+  47 Ops.WHERE           : 1          dtypes.float                             [41, 46, 6]                      None
+  48 Ops.CMPLT           : 1          dtypes.bool                              [47, 6]                          None
   49 Ops.CONST           :            dtypes.weakfloat                         []                               -0.5
   50 Ops.CAST            :            dtypes.float                             ['-0.5']                         dtypes.float
   51 Ops.CONST           :            dtypes.weakfloat                         []                               0.5
@@ -1015,7 +1053,7 @@
   79 Ops.CONST           :            dtypes.weakfloat                         []                               1.0
   80 Ops.CAST            :            dtypes.float                             ['1.0']                          dtypes.float
   81 Ops.ADD             : 1          dtypes.float                             [78, 80]                         None
-  82 Ops.SHR             : 1          dtypes.int                               [55, 7]                          None
+  82 Ops.SHR             : 1          dtypes.int                               [55, 9]                          None
   83 Ops.CONST           :            dtypes.weakint                           []                               127
   84 Ops.CAST            :            dtypes.int                               ['127']                          dtypes.int
   85 Ops.ADD             : 1          dtypes.int                               [82, 84]                         None
@@ -1030,55 +1068,56 @@
   94 Ops.BITCAST         : 1          dtypes.float                             [93]                             dtypes.float
   95 Ops.MUL             : 1          dtypes.float                             [90, 94]                         None
   96 Ops.WHERE           : 1          dtypes.float                             [36, 38, 95]                     None
-  97 Ops.WHERE           : 1          dtypes.float                             [28, 30, 96]                     None
-  98 Ops.WHERE           : 1          dtypes.float                             [23, 25, 97]                     None
-  99 Ops.INDEX           : 1          dtypes.float                             [12, 7]                          None
- 100 Ops.SUB             : 1          dtypes.float                             [99, 18]                         None
- 101 Ops.MUL             : 1          dtypes.float                             [100, 21]                        None
- 102 Ops.CMPNE           : 1          dtypes.bool                              [101, 101]                       None
- 103 Ops.CMPLT           : 1          dtypes.bool                              [101, 27]                        None
- 104 Ops.CMPLT           : 1          dtypes.bool                              [101, 32]                        None
- 105 Ops.CMPNE           : 1          dtypes.bool                              [104, 35]                        None
- 106 Ops.CMPNE           : 1          dtypes.bool                              [101, 38]                        None
- 107 Ops.CMPNE           : 1          dtypes.bool                              [101, 43]                        None
- 108 Ops.WHERE           : 1          dtypes.float                             [107, 101, 30]                   None
- 109 Ops.WHERE           : 1          dtypes.float                             [102, 30, 108]                   None
- 110 Ops.WHERE           : 1          dtypes.float                             [106, 109, 30]                   None
- 111 Ops.CMPLT           : 1          dtypes.bool                              [110, 30]                        None
- 112 Ops.WHERE           : 1          dtypes.float                             [111, 50, 52]                    None
- 113 Ops.ADD             : 1          dtypes.float                             [110, 112]                       None
- 114 Ops.CAST            : 1          dtypes.int                               [113]                            dtypes.int
- 115 Ops.CAST            : 1          dtypes.float                             [114]                            dtypes.float
- 116 Ops.SUB             : 1          dtypes.float                             [110, 115]                       None
- 117 Ops.MUL             : 1          dtypes.float                             [40, 116]                        None
- 118 Ops.ADD             : 1          dtypes.float                             [117, 60]                        None
- 119 Ops.MUL             : 1          dtypes.float                             [118, 116]                       None
- 120 Ops.ADD             : 1          dtypes.float                             [119, 64]                        None
- 121 Ops.MUL             : 1          dtypes.float                             [120, 116]                       None
- 122 Ops.ADD             : 1          dtypes.float                             [121, 68]                        None
- 123 Ops.MUL             : 1          dtypes.float                             [122, 116]                       None
- 124 Ops.ADD             : 1          dtypes.float                             [123, 72]                        None
- 125 Ops.MUL             : 1          dtypes.float                             [124, 116]                       None
- 126 Ops.ADD             : 1          dtypes.float                             [125, 76]                        None
- 127 Ops.MUL             : 1          dtypes.float                             [126, 116]                       None
- 128 Ops.ADD             : 1          dtypes.float                             [127, 80]                        None
- 129 Ops.SHR             : 1          dtypes.int                               [114, 7]                         None
- 130 Ops.ADD             : 1          dtypes.int                               [129, 84]                        None
- 131 Ops.SHL             : 1          dtypes.int                               [130, 87]                        None
- 132 Ops.BITCAST         : 1          dtypes.float                             [131]                            dtypes.float
- 133 Ops.MUL             : 1          dtypes.float                             [128, 132]                       None
- 134 Ops.SUB             : 1          dtypes.int                               [114, 129]                       None
- 135 Ops.ADD             : 1          dtypes.int                               [134, 84]                        None
- 136 Ops.SHL             : 1          dtypes.int                               [135, 87]                        None
- 137 Ops.BITCAST         : 1          dtypes.float                             [136]                            dtypes.float
- 138 Ops.MUL             : 1          dtypes.float                             [133, 137]                       None
- 139 Ops.WHERE           : 1          dtypes.float                             [105, 38, 138]                   None
- 140 Ops.WHERE           : 1          dtypes.float                             [103, 30, 139]                   None
- 141 Ops.WHERE           : 1          dtypes.float                             [102, 25, 140]                   None
- 142 Ops.ADD             : 1          dtypes.float                             [98, 141]                        None
- 143 Ops.STORE           : 1          dtypes.void                              [4, 142]                         None
- 144 Ops.END             :            dtypes.void                              [143, 3]                         None
- 145 Ops.SINK            :            dtypes.void                              [144]                            KernelInfo(name='r_3_2', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)),), opts_to_apply=None, estimates=None, beam=0)
+  97 Ops.WHERE           : 1          dtypes.float                             [30, 6, 96]                      None
+  98 Ops.WHERE           : 1          dtypes.float                             [25, 27, 97]                     None
+  99 Ops.ADD             : 1          dtypes.float                             [6, 98]                          None
+ 100 Ops.INDEX           : 1          dtypes.float                             [14, 9]                          None
+ 101 Ops.SUB             : 1          dtypes.float                             [100, 20]                        None
+ 102 Ops.MUL             : 1          dtypes.float                             [101, 23]                        None
+ 103 Ops.CMPNE           : 1          dtypes.bool                              [102, 102]                       None
+ 104 Ops.CMPLT           : 1          dtypes.bool                              [102, 29]                        None
+ 105 Ops.CMPLT           : 1          dtypes.bool                              [102, 32]                        None
+ 106 Ops.CMPNE           : 1          dtypes.bool                              [105, 35]                        None
+ 107 Ops.CMPNE           : 1          dtypes.bool                              [102, 38]                        None
+ 108 Ops.CMPNE           : 1          dtypes.bool                              [102, 43]                        None
+ 109 Ops.WHERE           : 1          dtypes.float                             [108, 102, 6]                    None
+ 110 Ops.WHERE           : 1          dtypes.float                             [103, 6, 109]                    None
+ 111 Ops.WHERE           : 1          dtypes.float                             [107, 110, 6]                    None
+ 112 Ops.CMPLT           : 1          dtypes.bool                              [111, 6]                         None
+ 113 Ops.WHERE           : 1          dtypes.float                             [112, 50, 52]                    None
+ 114 Ops.ADD             : 1          dtypes.float                             [111, 113]                       None
+ 115 Ops.CAST            : 1          dtypes.int                               [114]                            dtypes.int
+ 116 Ops.CAST            : 1          dtypes.float                             [115]                            dtypes.float
+ 117 Ops.SUB             : 1          dtypes.float                             [111, 116]                       None
+ 118 Ops.MUL             : 1          dtypes.float                             [40, 117]                        None
+ 119 Ops.ADD             : 1          dtypes.float                             [118, 60]                        None
+ 120 Ops.MUL             : 1          dtypes.float                             [119, 117]                       None
+ 121 Ops.ADD             : 1          dtypes.float                             [120, 64]                        None
+ 122 Ops.MUL             : 1          dtypes.float                             [121, 117]                       None
+ 123 Ops.ADD             : 1          dtypes.float                             [122, 68]                        None
+ 124 Ops.MUL             : 1          dtypes.float                             [123, 117]                       None
+ 125 Ops.ADD             : 1          dtypes.float                             [124, 72]                        None
+ 126 Ops.MUL             : 1          dtypes.float                             [125, 117]                       None
+ 127 Ops.ADD             : 1          dtypes.float                             [126, 76]                        None
+ 128 Ops.MUL             : 1          dtypes.float                             [127, 117]                       None
+ 129 Ops.ADD             : 1          dtypes.float                             [128, 80]                        None
+ 130 Ops.SHR             : 1          dtypes.int                               [115, 9]                         None
+ 131 Ops.ADD             : 1          dtypes.int                               [130, 84]                        None
+ 132 Ops.SHL             : 1          dtypes.int                               [131, 87]                        None
+ 133 Ops.BITCAST         : 1          dtypes.float                             [132]                            dtypes.float
+ 134 Ops.MUL             : 1          dtypes.float                             [129, 133]                       None
+ 135 Ops.SUB             : 1          dtypes.int                               [115, 130]                       None
+ 136 Ops.ADD             : 1          dtypes.int                               [135, 84]                        None
+ 137 Ops.SHL             : 1          dtypes.int                               [136, 87]                        None
+ 138 Ops.BITCAST         : 1          dtypes.float                             [137]                            dtypes.float
+ 139 Ops.MUL             : 1          dtypes.float                             [134, 138]                       None
+ 140 Ops.WHERE           : 1          dtypes.float                             [106, 38, 139]                   None
+ 141 Ops.WHERE           : 1          dtypes.float                             [104, 6, 140]                    None
+ 142 Ops.WHERE           : 1          dtypes.float                             [103, 27, 141]                   None
+ 143 Ops.ADD             : 1          dtypes.float                             [99, 142]                        None
+ 144 Ops.STORE           : 1          dtypes.void                              [4, 143]                         None
+ 145 Ops.END             :            dtypes.void                              [144, 3]                         None
+ 146 Ops.SINK            :            dtypes.void                              [145]                            KernelInfo(name='r_3_2', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)),), opts_to_apply=None, estimates=None, beam=0)
 === kernel 13 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 6, device='CPU')
    1 Ops.CONST           :            dtypes.weakint                           []                               3
